sql: fix index column direction lookup bugs

postamar opened the pull request:
This patch fixes a couple of instances in which we look up an index's
column directions using a wrong ordinal when the index has implicit
columns due to it being sharded or partitioned.

Fixes #58945.

Release note (bug fix): The indoption column in pg_catalog.index is now
populated correctly.

postamar commented on pkg/sql/catalog/tabledesc/index.go:
```diff
@@ -273,6 +273,11 @@ func (w index) ForEachColumnID(fn func(colID descpb.ColumnID) error) error {
 
+// FirstExplicitColumnOrdinal returns the first ordinal of any explicit columns.
+func (w index) FirstExplicitColumnOrdinal() int {
+	return w.desc.FirstExplicitColumnOrdinal()
```
This is in line with my recent changes from https://github.com/cockroachdb/cockroach/pull/58678 and the method renaming was done for the sake of naming consistency with the rest of `catalog.Index`.

rafiss reviewed (changes requested):
thanks for picking this up! have a few clarifying questions for my understanding

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-:-MQxzSML6M2Ch0YdDDFQ:b-d6ev3l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[pkg/sql/crdb_internal.go, line 2123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-MQxuwm_1wTTEEacny8_:-MQxuwmaE85WP0L6kMJ_:b-qpwmk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aac346fc42322c9c600afcaea8687bf58ebed38/pkg/sql/crdb_internal.go#L2123)):*
> ```Go
> 						colName := tree.DNull
> 						colDir := tree.DNull
> 						if i >= len(index.IndexDesc().ColumnNames) {
> ```

nit: it seems like it would be a bit more readable if `index.IndexDesc()` were extracted into a separate variable (as line 2115 was doing before) any particular reason not to?
___
*[pkg/sql/pg_catalog.go, line 1453 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-MQxzIBO0i-qQK2IsDXr:-MQxzIBO0i-qQK2IsDXs:b-rjwj25) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aac346fc42322c9c600afcaea8687bf58ebed38/pkg/sql/pg_catalog.go#L1453)):*
> ```Go
> 					indoption := tree.NewDArray(types.Int)
> 					colIDs := make([]descpb.ColumnID, 0, index.NumColumns())
> 					for i := index.FirstExplicitColumnOrdinal(); i < index.NumColumns(); i++ {
> ```

actually i'm still confused what the original bug was. is the old `ExplicitColumnStartIdx()` not the same as the new `FirstExplicitColumnOrdinal()` ?
___
*[pkg/sql/pg_catalog.go, line 1496 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-MQxw6QNAGkQJYG0v2AK:-MQxw6QNAGkQJYG0v2AL:b-hrnwkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aac346fc42322c9c600afcaea8687bf58ebed38/pkg/sql/pg_catalog.go#L1496)):*
> ```Go
> 						tree.MakeDBool(tree.DBool(index.IsUnique())),           // indimmediate
> 						tree.DBoolFalse,                                        // indisclustered
> 						tree.MakeDBool(tree.DBool(index.Public())),             // indisvalid
> ```

i'm not so familiar with this. why is `index.Public()` the same as `!isMutation`
___
*[pkg/sql/pg_catalog.go, line 1498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-MQxwVgW0btNDIdfKbNp:-MQxwVgW0btNDIdfKbNq:b-gd0rsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aac346fc42322c9c600afcaea8687bf58ebed38/pkg/sql/pg_catalog.go#L1498)):*
> ```Go
> 						tree.MakeDBool(tree.DBool(index.Public())),             // indisvalid
> 						tree.DBoolFalse,                                        // indcheckxmin
> 						tree.MakeDBool(tree.DBool(index.WriteAndDeleteOnly())), // indisready
> ```

similar question, why is `index.WriteAndDeleteOnly()` equivalent?
___
*[pkg/sql/pg_catalog.go, line 1549 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-MQxykQjBifTAzgjCSBS:-MQxykQjBifTAzgjCSBT:bkivhu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aac346fc42322c9c600afcaea8687bf58ebed38/pkg/sql/pg_catalog.go#L1549)):*
> ```Go
> 	db *dbdesc.Immutable,
> 	table catalog.TableDescriptor,
> 	index catalog.Index,
> ```

should the function be renamed? it no longer seems to be "from a descriptor"
___
*[pkg/sql/catalog/tabledesc/index.go, line 278 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-MQwfT-sad1WjNd59Edb-r1-278:-MQxuR2S31ClwJz3DiCL:b6ielg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aac346fc42322c9c600afcaea8687bf58ebed38/pkg/sql/catalog/tabledesc/index.go#L278)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is in line with my recent changes from https://github.com/cockroachdb/cockroach/pull/58678 and the method renaming was done for the sake of naming consistency with the rest of `catalog.Index`.
</blockquote></details>

sounds good! a small request for the future: can you separate any renames/refactors into a commit separate from the commit that has functionality changes (and include both in the same PR). it helps with reviewing. definitely not a big deal for this PR, since the changes are small


<!-- Sent from Reviewable.io -->

Commit:
sql: fix index column direction lookup bugs

This patch fixes a couple of instances in which we look up an index's
column directions using a wrong ordinal when the index has implicit
columns due to it being sharded or partitioned.

Fixes #58945.

Release note (bug fix): The indoption column in pg_catalog.index is now
populated correctly.

postamar reviewed (commented):
Thanks for the review. You're right there are too many unrelated changes in this PR. Here it is pared down to the bug fix, RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-:-MR0Hzyl4_fQ0-x91J8L:byow0kt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)



<!-- Sent from Reviewable.io -->

postamar commented:
CI failed due to an unrelated test failure for which there is already an issue open: https://github.com/cockroachdb/cockroach/issues/58989

rafiss reviewed (approved):
thanks for the cleanup! i understand what's going on now :)

if you'd like to include your other refactoring, feel free to add that as another commit in this PR. if not, this lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58947#-:-MR1fpVoBur-pFNWTMhw:bis85s5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)



<!-- Sent from Reviewable.io -->

rafiss commented:
i've re-run CI to see if it doesn't flake this time

postamar commented:
bors r+

Thanks, I think I'll open another PR for the rest of the changes, they're basically loose ends from https://github.com/cockroachdb/cockroach/pull/58678 , so I expect I'll find more of them.
